light-prover/prover/batch_append_circuit.go

```diff
@@ -0,0 +1,313 @@
+)
+
+type BatchAppendCircuit struct {
+	// Public inputs
+	OldSubTreeHashChain frontend.Variable `gnark:",public"`
+	NewSubTreeHashChain frontend.Variable `gnark:",public"`
+	NewRoot             frontend.Variable `gnark:",public"`
+	HashchainHash       frontend.Variable `gnark:",public"`
+	StartIndex          frontend.Variable `gnark:",public"`
+
+	// Private inputs
+	Leaves              []frontend.Variable `gnark:",private"`
+	Subtrees            []frontend.Variable `gnark:",private"`
+	HashChainStartIndex frontend.Variable   `gnark:",private"`
+
+	BatchSize uint32
+	Depth     uint32
+}
+
+func (circuit *BatchAppendCircuit) Define(api frontend.API) error {
+	if len(circuit.Leaves) != int(circuit.BatchSize) {
+		return fmt.Errorf("leaves length (%d) does not match batch size (%d)", len(circuit.Leaves), circuit.BatchSize)
+	}
+
+	// Assert HashChainStartIndex is equal to StartIndex
+	api.AssertIsEqual(circuit.HashChainStartIndex, circuit.StartIndex)
+
+	// Validate Subtrees
+	subtreesHashChain := circuit.validateSubtrees(api)
+	api.AssertIsEqual(subtreesHashChain, circuit.OldSubTreeHashChain)
+
+	// Validate leaves hash chain
+	leavesHashChain := circuit.validateLeavesHashChain(api)
+	api.AssertIsEqual(leavesHashChain, circuit.HashchainHash)
+
+	// Append leaves
+	newRoot, newSubtrees := circuit.batchAppend(api)
+
+	// Assert new root
+	api.AssertIsEqual(newRoot, circuit.NewRoot)
+
+	// Create and validate new subtrees hash chain
+	newSubtreesHashChain := circuit.createSubtreesHashChain(api, newSubtrees)
+	api.AssertIsEqual(newSubtreesHashChain, circuit.NewSubTreeHashChain)
+
+	return nil
+}
+
+func (circuit *BatchAppendCircuit) validateSubtrees(api frontend.API) frontend.Variable {
+	hashChain := circuit.OldSubTreeHashChain
+	for i := 0; i < int(circuit.Depth); i++ {
+		hashChain = abstractor.Call(api, poseidon.Poseidon2{In1: hashChain, In2: circuit.Subtrees[i]})
+	}
+	return hashChain
+}
+
+func (circuit *BatchAppendCircuit) validateLeavesHashChain(api frontend.API) frontend.Variable {
+	hashChain := frontend.Variable(0)
+	for _, leaf := range circuit.Leaves {
+		hashChain = abstractor.Call(api, poseidon.Poseidon2{In1: hashChain, In2: leaf})
+	}
+	return hashChain
+}
+
+func (circuit *BatchAppendCircuit) batchAppend(api frontend.API) (frontend.Variable, []frontend.Variable) {
+	currentRoot := circuit.OldSubTreeHashChain
+	currentSubtrees := make([]frontend.Variable, len(circuit.Subtrees))
+	copy(currentSubtrees, circuit.Subtrees)
+
+	for i := 0; i < int(circuit.BatchSize); i++ {
+		index := api.Add(circuit.StartIndex, i)
+		leaf := circuit.Leaves[i]
+		currentRoot, currentSubtrees = circuit.append(api, leaf, currentSubtrees, index)
+	}
+
+	return currentRoot, currentSubtrees
+}
+
+func (circuit *BatchAppendCircuit) append(api frontend.API, leaf frontend.Variable, subtrees []frontend.Variable, index frontend.Variable) (frontend.Variable, []frontend.Variable) {
+	path := api.ToBinary(index, int(circuit.Depth))
```

Comment:
an optimization we can look into once `append` and `update` is merged

`ToBinary` is quite expensive we could try optimize this:
- remove `ToBinary` here
- only call `ToBinary` once per proof
- increment the binary array directly